GoldSrc voice chat

rootcreep opened the pull request:
## Voice Packet Structure

| Offset (bytes) | Length (bytes) | Field Name           | Description                                                    |
|----------------|---------------|----------------------|----------------------------------------------------------------|
| 0              | 8             | SteamID              | Unique user identifier (Steam)                                 |
| 8              | 1             | VPC_SETSAMPLERATE    | Command type: always 11                                        |
| 9              | 2             | Sample Rate          | Sample rate, Little Endian (e.g., 24000)                       |
| 11             | 1             | VPC_VDATA_OPUS_PLC   | Command type: [see code](https://github.com/FWGS/xash3d-fwgs/blob/cca1732245a6d90a13a14b05859d4132d3fe60ab/engine/client/voice.h#L54)                     |
| 12             | 2             | Data Length          | Length of the following data (Opus or Silence), Little Endian  |
| 14             | N             | Data                 | Opus data (N bytes) or silence (N samples)                     |
| 14+N           | 4             | CRC32                | CRC32 checksum of all previous bytes in the packet             |

**Notes:**
- If **VPC_VDATA_OPUS_PLC** (6), the following N bytes are Opus-encoded audio data.
- If **VPC_VDATA_SILENCE** (0), the following N samples are silence (usually just zeros).
- Sample Rate is almost always 24000.
- CRC32 is calculated over all bytes except the last 4 (the CRC32 itself).

## Opus Payload Structure

The Opus data (payload) inside a voice packet consists of a sequence of Opus frames, each with the following format:

| Offset (from start of Opus payload) | Length (bytes) | Field Name   | Description                                                        |
|-------------------------------------|---------------|--------------|--------------------------------------------------------------------|
| 0                                   | 2             | frame_size   | Size of the Opus frame in bytes (uint16, little-endian)            |
| 2                                   | 2             | sequence     | Sequence number/counter (uint16, little-endian), not used in code  |
| 4                                   | N             | frame_data   | Opus-encoded data, N bytes                                         |

Then the next frame follows (if any), and so on until the end of the Opus payload.

### Special values for frame_size:
- **0** — means "silence" (add 160 zero samples to the output buffer).
- **0xFFFF** — Opus decoder reset command; after this, parsing of the Opus payload stops.

### Example

| Offset | Length | Value    | Description                  |
|--------|--------|----------|------------------------------|
| 0      | 2      | 0x0032   | frame_size = 50 bytes        |
| 2      | 2      | 0x0001   | sequence = 1                 |
| 4      | 50     | ...      | Opus data (50 bytes)         |
| 54     | 2      | 0x0000   | frame_size = 0 (silence)     |
| 56     | 2      | 0x0002   | sequence = 2                 |
| 58     | 0      | -        | no data, just 160 zeros      |

rootcreep committed:
engine: sdl: always reset device descriptors on shutdown

rootcreep committed:
engine: sound: voice buffer isolation and correct MP3 volume handling

rootcreep committed:
engine: client: GoldSrc voice chat support

- Added support for GoldSrc voice mode, including initialization and shutdown functions.
- Implemented data size verification for the GoldSrc voice package.
- Creation and processing of voice packets, CRC verification of data integrity.
- Added a Steam ID field to the client_static_t structure
- Automatic gain control has been introduced for voice samples.
- The voice status structure has been updated.
- Refactored voice data handling for better type consistency and maintainability.
- Fixed ISO C90 compliance issues for better portability.
- Fixed decoder index usage.

a1batross commented on engine/client/voice.c:
```diff
@@ -249,63 +510,274 @@ static uint Voice_GetOpusCompressedData( byte *out, uint maxsize, uint *frames )
+
+        int is_silence = 1;
+        int16_t *samples = (int16_t *)(voice.input_buffer + ofs);
+        for (uint i = 0; i < frame_size_samples; ++i) {
```
Again, you can just reformat whole file using uncrustify. It's very easy to use from console. Grab the _latest_ version and run:

`uncrustify -c uncrustify.cfg --replace --no-backup engine/client/voice.c`

a1batross commented on engine/client/voice.c:
```diff
@@ -249,63 +510,274 @@ static uint Voice_GetOpusCompressedData( byte *out, uint maxsize, uint *frames )
+        return 0;
+
+    sample_rate = LittleShort(*(uint16_t*)(data + offset));
+    offset += 2;
```
Use `sizeof( uint16_t )` instead.

a1batross commented on engine/client/voice.c:
```diff
@@ -249,63 +510,274 @@ static uint Voice_GetOpusCompressedData( byte *out, uint maxsize, uint *frames )
+			Voice_ApplyGainAdjust(samples, frame_size_samples, voice_transmit_scale.value);
+
+        bytes = opus_encode( voice.gs_encoder, samples,
+            frame_size_samples, out + size + 4, maxsize - 4 );
```
Same thing about magic numbers here.

If `4` represents 32-bit number, use `sizeof( uint32_t )`. If it represents two separate 16-bit values, use `sizeof( uint16_t ) * 2` to keep the code readable and maintainable.

a1batross commented on engine/client/voice.c:
```diff
@@ -249,63 +510,274 @@ static uint Voice_GetOpusCompressedData( byte *out, uint maxsize, uint *frames )
+            opus_offset += 4;
+
+            if (frame_size == 0) {
+                if (output_samples + 160 > GS_MAX_DECOMPRESSED_SAMPLES) {
```
What does `160` is supposed to mean here?

a1batross commented on engine/client/voice.c:
```diff
@@ -585,14 +1108,26 @@ static void Voice_Shutdown( void )
 	}
 
-	memset( &voice, 0, sizeof( voice ));
+	voice.initialized = false;
```
Why memset is gone?

a1batross commented on engine/client/voice.c:
```diff
@@ -605,15 +1140,19 @@ void Voice_Idle( double frametime )
 	// update local player status first
-	Voice_StatusTimeout( &voice.local, VOICE_LOOPBACK_INDEX, frametime );
+	if ( voice.goldsrc )
+		Voice_StatusTimeout( &voice.local, cl.playernum + 1, frametime );
```
What so special about goldsrc protocol when notifying client.dll?

a1batross commented on engine/client/voice.c:
```diff
@@ -628,67 +1167,53 @@ Initialize the voice subsystem
+	{
+		// unsupported codec
+		Voice_Shutdown();
+		voice.goldsrc = false;
```
Why? If the voice hasn't been initialized, cleaning up the whole structure is much more fool-proof.

Buggem commented on engine/client/voice.c:
```diff
@@ -628,67 +1167,53 @@ Initialize the voice subsystem
 			return false;
 		}
-
-		// we can hear others players, so it's fine to fail now
```
Please keep these comments, it's easier for people to understand how this code works with them

rootcreep committed:
engine: client: replaced hardcoded buffer sizes with defined constants for voice data

rootcreep committed:
engine: client: refactored voice.c

- reformatted code.
- changed function declarations to static.
- ensured consistent use of defined constants for buffer sizes.

rootcreep committed:
engine: client: move voice reset to Voice_Shutdown
